app/scripts/views/settings.js

```diff
@@ -3,114 +3,296 @@
+      if (self._currentSubView && self._currentSubView.isModal) {
+        self._currentSubView.closePanel();
+      }
+
+      return self._createSubViewIfNeeded(SubView, options)
+        .then(function (subView) {
+          if (subView) {
+            self._currentSubView = subView;
+            subView.openPanel();
+            subView.logScreen();
+            return subView;
+          }
+        });
+    },
+
+    _subviewInstanceFromClass: function (SubView) {
+      return this.subviews.filter(function (subView) {
+        if (subView instanceof SubView) {
+          return true;
+        }
+      })[0];
+    },
+
+    _isModalViewClass: function (SubView) {
+      return !! SubView.prototype.isModal;
     },
 
-    submit: function () {
+    _subViewClass: function (SubView) {
+      return SubView.prototype.className;
+    },
+
+    // Render subview if an instance doesn't already exist
+    _createSubViewIfNeeded: function (SubView) {
+      var subView = this._subviewInstanceFromClass(SubView);
+      if (subView) {
+        return p(subView);
+      }
+
       var self = this;
-      var sessionToken = self.getSignedInAccount().get('sessionToken');
+      var className = self._subViewClass(SubView);
+      var selector = '.' + className;
 
-      self.logScreenEvent('signout.submit');
-      return self.fxaClient.signOut(sessionToken)
-        .fail(function () {
-          // ignore the error.
-          // Clear the session, even on failure. Everything is A-OK.
-          // See issue #616
-          self.logScreenEvent('signout.error');
-        })
-        .fin(function () {
-          self.logScreenEvent('signout.success');
-          self.user.clearSignedInAccount();
-          Session.clear();
-          self.navigate('signin', {
-            success: t('Signed out successfully')
-          });
+      self.$('#subviews').append('<div class="settings-subview ' + className + '"></div>');
+
+      var view = self.router.createView(SubView, {
+        el: self.$(selector),
+        superView: self
+      });
+
+      self.trackSubview(view);
+
+      return renderView(view);
+    },
+
+    beforeRender: function () {
+      var self = this;
+      $('body').addClass('settings');
+      var account = self.getSignedInAccount();
+
+      return account.fetchProfile()
+        .then(function () {
+          self.user.setAccount(account);
         });
     },
 
+    afterRender: function () {
+      var self = this;
+      var communicationPrefsAreVisible = self._areCommunicationPrefsVisible();
+
+      // Initial subviews to render; excludes CommunicationPreferencesView if not visible
+      // and modal views.
+      var initialSubViews = self._panelViews.filter(function (SubView) {
+        var isCommView = SubView === CommunicationPreferencesView;
+        var isModalView = self._isModalViewClass(SubView);
+        var shouldHide = (! communicationPrefsAreVisible && isCommView) || isModalView;
```

Comment:
I find this logic a bit hard to decipher. Can you serialize the logic a bit, and maybe change from negative to positive tests?

```js
if (SubView === CommunicationPreferencesView) {
  return areCommunicationPrefsVisible;
} else {
  return ! isModalView;
}
```